editoast/src/views/infra/delimited_area.rs

```diff
@@ -0,0 +1,837 @@
+            track_exits.push(exit);
+        } else {
+            let track_exits = vec![exit];
+            tracks_to_exits.insert(track_id, track_exits);
+        };
+    });
+    let exits = tracks_to_exits;
+
+    // Directional track ranges reachable from `entry` during the graph exploration.
+    let mut related_track_ranges: Vec<DirectionalTrackRange> = Vec::new();
+
+    // TrackEndpoint right after the entry location (in the correct direction):
+    let first_track_endpoint = TrackEndpoint {
+        endpoint: match entry.direction {
+            Direction::StartToStop => Endpoint::End,
+            Direction::StopToStart => Endpoint::Begin,
+        },
+        track: entry.track.clone(),
+    };
+
+    if let Some(immediate_exit) = closest_exit_from_entry(entry, exits.get(&entry.track)) {
+        if let Some(only_track_range) = track_range_between_two_locations(entry, immediate_exit) {
+            return vec![only_track_range];
+        } else {
+            return vec![];
+        }
+    } else if let Some(first_track_range) =
+        track_range_between_endpoint_and_location(entry, &first_track_endpoint, infra_cache, true)
+    {
+        related_track_ranges.push(first_track_range);
+    } else {
+        return vec![];
+    };
+
+    // Identifiers of the track sections that have already been reached and should be ignored:
+    let mut visited_tracks: HashSet<&TrackEndpoint> = HashSet::new();
+
+    // Neighbors of the explored tracks, i.e. the tracks that should be visited next:
+    let mut next_tracks: Vec<(&TrackEndpoint, f64)> = Vec::new();
+    let remaining_distance =
+        max_distance - (related_track_ranges[0].end - related_track_ranges[0].begin);
+    if 0. < remaining_distance {
+        let neighbours = graph
+            .get_all_neighbours(&first_track_endpoint)
+            .into_iter()
+            .map(|neighbour| (neighbour, remaining_distance))
+            .collect::<Vec<_>>();
+        next_tracks.extend(neighbours);
+    }
+
+    while let Some((curr_track_endpoint, remaining_distance)) = next_tracks.pop() {
+        let curr_track_id = &curr_track_endpoint.track;
+
+        if !visited_tracks.insert(curr_track_endpoint) {
+            // Track already visited
+            continue;
+        }
+
+        // Check if there is an exit location on that track range
+        if let Some(exit) =
+            closest_exit_from_endpoint(curr_track_endpoint, exits.get(&curr_track_id))
+        {
+            // End the search on that track, add the current track with the correct offset
+            let track_range = track_range_between_endpoint_and_location(
+                exit,
+                curr_track_endpoint,
+                infra_cache,
+                false,
+            )
+            .unwrap_or_else(|| {
+                panic!(
+                    "Cannot form a track from {:?} and {:?}",
+                    exit, curr_track_endpoint
+                )
```

Comment:
At this point, we have already retrieved the track range that corresponds to the entry signal given by the user. The track ranges being explored are retrieved from the loaded infra graph. If we cannot form a track range inside the main loop, it is either caused by a bug in the implementation or by invalid data in the loaded graph. In both cases, wouldn't a 5xx error be more appropriate ?

Reply:
I think it is great. I updated the API so that it returns a list of invalid user inputs and their associated error in case of failure and kept the asserts in the graph exploration. Some utility methods were updated so that all asserts happen at the same level: directly in the main `impacted_tracks` method instead of having some failures like trying to build a track with invalid values panic in the utility functions that build the track ranges (for example, if the infra cache did not contain a matching track for an identifier).

Overall we have now:
* A `DelimitedAreaError` that handles http responses in case of invalid user input.
* A `TrackRangeConstructionError` that is used to catch internal errors which have nothing to do with the API.
* panics in the main graph exploration loop when falling on a supposedly impossible situation (bug in the implementation or incorrect data in the cache).

What do you think ?

Reply:
That's great. I believe it's much more readable like this :)